File: managed/api/GrapeEngine.Scripting/Job/ThreadAffinity.cs
/* Start Header *****************************************************************/
/*!
\file   ThreadAffinity.cs
\author Muhammad Nur Fadzly Bin Zulkifli (100%)
\par    muhammadnurfadzly.b@digipen.edu
\brief
Managed wrapper for thread affinity utilities - bind threads to CPU cores
for better cache locality and performance.

Copyright (C) 2025 DigiPen Institute of Technology.
Reproduction or disclosure of this file or its contents without the
prior written consent of DigiPen Institute of Technology is prohibited.
*/
/* End Header *******************************************************************/

using GrapeEngine.Scripting.Unsafe;

namespace GrapeEngine.Scripting.Job;

/// <summary>
/// Utilities for managing thread affinity.
/// 
/// Thread affinity binds a thread to specific CPU cores, which can improve
/// performance by keeping thread data in the core's cache and reducing
/// context switching overhead.
/// 
/// Thread affinity is automatically configured by the JobManager, but this
/// class can be used for custom thread management.
/// </summary>
public static class ThreadAffinity
{
    /// <summary>
    /// Check if thread affinity is supported on this platform.
    /// </summary>
    public static bool IsSupported => JobSystemAPI.ThreadAffinityIsSupported();

    /// <summary>
    /// Get the number of available CPU cores.
    /// </summary>
    public static uint NumCores => JobSystemAPI.ThreadAffinityGetNumCores();

    /// <summary>
    /// Bind the current thread to a specific CPU core.
    /// </summary>
    /// <param name="coreIndex">The core to bind to (0-based)</param>
    /// <returns>true if successful, false if not supported or invalid core</returns>
    public static bool SetCurrentThreadAffinity(uint coreIndex)
    {
        if (!IsSupported || coreIndex >= NumCores) return false;
        
        return JobSystemAPI.ThreadAffinitySetCurrentThreadAffinity(coreIndex);
    }

    /// <summary>
    /// Get suggested thread affinity for a worker thread index.
    /// </summary>
    /// <param name="threadIndex">The worker thread index</param>
    /// <returns>Suggested CPU core (wraps around if needed)</returns>
    public static uint GetSuggestedAffinity(uint threadIndex)
    {
        return threadIndex % NumCores;
    }
}

File: engine/src/ecs/jobs/ThreadAffinity.cpp
/* Start Header *****************************************************************/
/*!
\file    ThreadAffinity.cpp
\author  Muhammad Nur Fadzly Bin Zulkifli (100%)
\par     muhammadnurfadzly.b@digipen.edu
\brief
This file contains the implementation of ThreadAffinity utilities for setting
thread affinity on Windows and other platforms.

Copyright (C) 2025 DigiPen Institute of Technology.
Reproduction or disclosure of this file or its contents without the
prior written consent of DigiPen Institute of Technology is prohibited.
*/
/* End Header *******************************************************************/

#include "ecs/jobs/ThreadAffinity.h"
#include <thread>

#ifdef _WIN32
    #include <windows.h>
#endif

namespace ECS::Jobs {

    bool ThreadAffinity::SetCurrentThreadAffinity(uint32_t coreIndex) {
#ifdef _WIN32
        // Windows implementation using SetThreadAffinityMask
        HANDLE currentThread = GetCurrentThread();
        DWORD_PTR affinityMask = 1ULL << coreIndex;
        
        DWORD_PTR result = SetThreadAffinityMask(currentThread, affinityMask);
        return result != 0;
#else
        // Unsupported platform
        return false;
#endif
    }

    bool ThreadAffinity::SetThreadAffinity(std::thread::native_handle_type handle, 
                                          uint32_t coreIndex) {
#ifdef _WIN32
        // Windows implementation
        DWORD_PTR affinityMask = 1ULL << coreIndex;
        DWORD_PTR result = SetThreadAffinityMask(handle, affinityMask);
        return result != 0;
#else
        // Unsupported platform
        return false;
#endif
    }

    uint32_t ThreadAffinity::GetNumCores() {
        return std::thread::hardware_concurrency();
    }

    bool ThreadAffinity::IsSupported() {
#ifdef _WIN32
        return true;
#else
        return false;
#endif
    }

}
